R/importance.R

```diff
@@ -36,131 +39,243 @@
+        x_perm_name <- task$add_columns(x_perm)
+        task_x_perm <- task$next_in_chain(column_names = x_perm_name)
+        # obtain predictions & risk on the new task with permuted x
+        x_perm_pred <- fit$predict_fold(task_x_perm, fold_number)
+        x_perm_eval <- eval_fun(x_perm_pred, Y)
+        if (!is.null(attr(original_eval, "loss")) && !attr(original_eval, "loss")) {
+          no_x_risk <- x_perm_eval
+        } else {
+          no_x_losses <- x_perm_eval
+          no_x_risk <- mean(no_x_losses)
+        }
+        return(no_x_risk)
+      }, mc.cores = cores)))
     } else if (type == "remove") {
-      # modify learner to not include covariate x
-      x_removed_lrnr <- fit$reparameterize(list(covariates = setdiff(X, x)))
-      x_removed_fit <- x_removed_lrnr$train(task)
-      x_removed_pred <- x_removed_fit$predict_fold(task, fold_number)
-      no_x_risk <- mean(loss(x_removed_pred, Y))
+      # modify learner to not include x (covariate/group)
+      if (!is.null(covariate_groups)) {
+        x_rm_covars <- setdiff(unlist(X), x)
+      } else {
+        x_rm_covars <- setdiff(X, x)
+      }
+      x_rm_lrnr <- fit$reparameterize(list(covariates = x_rm_covars))
+      x_rm_fit <- x_rm_lrnr$train(task)
+      x_rm_pred <- x_rm_fit$predict_fold(task, fold_number)
+      x_rm_eval <- eval_fun(x_rm_pred, Y)
+      if (!is.null(attr(original_eval, "loss")) && !attr(original_eval, "loss")) {
+        no_x_risk <- x_rm_eval
+      } else {
+        no_x_losses <- x_rm_eval
+        no_x_risk <- mean(no_x_losses)
+      }
     }
+
     # evaluate importance
     if (importance_metric == "ratio") {
-      return(no_x_risk / original_risk)
+      result <- no_x_risk / original_risk
     } else if (importance_metric == "difference") {
-      return(no_x_risk - original_risk)
+      result <- no_x_risk - original_risk
+    }
+
+    if (type == "permute") {
+      result <- list("result" = result, "RNGseeds" = seeds) # reproducibility
+    } else if (type == "remove") {
+      result <- list("result" = result)
     }
+
+    return(result)
   })
-  names(res_list) <- X
 
+  if (type == "permute") {
+    RNGseeds <- data.table(do.call(cbind, lapply(res_list, "[[", "RNGseeds")))
+    colnames(RNGseeds) <- names(X)
+  }
+  res_list <- lapply(res_list, "[[", "result")
+
+  ############################## prep output ###################################
   # importance results ordered by decreasing importance
-  if (importance_metric == "ratio") {
-    res <- data.table(X = names(res_list), risk_ratio = unlist(res_list))
-    return(res[order(-res$risk_ratio)])
-  } else if (importance_metric == "difference") {
-    res <- data.table(X = names(res_list), risk_difference = unlist(res_list))
-    return(res[order(-res$risk_difference)])
+  result <- data.table(covariate = names(X), metric = unlist(res_list))
+  if (!is.null(covariate_groups)) {
+    colnames(result)[1] <- "covariate_group"
+  }
+  result <- result[order(-result$metric), ]
```

Comment:
this is not strictly equivalent to `data.table::setorder()` (https://www.rdocumentation.org/packages/data.table/versions/1.14.0/topics/setorder) even though `result` is a `data.table`. probably `result` is not a large `data.table` such that re-ordering would be memory inefficient by why not bother to do this by reference since you're already using `data.table`?